capsules/src/net/lowpan_fragment.rs

```
@@ -0,0 +1,900 @@
+        });
+    }
+
+    fn receive_frame(&self,
+                     packet: &[u8],
+                     packet_len: usize,
+                     src_mac_addr: MacAddress,
+                     dst_mac_addr: MacAddress)
+                     -> (Option<&RxState<'a>>, ReturnCode) {
+        if is_fragment(packet) {
+            let (is_frag1, dgram_size, dgram_tag, dgram_offset) = get_frag_hdr(&packet[0..5]);
+            let offset_to_payload = if is_frag1 {
+                lowpan_frag::FRAG1_HDR_SIZE
+            } else {
+                lowpan_frag::FRAGN_HDR_SIZE
+            };
+            self.receive_fragment(&packet[offset_to_payload..],
+                                  packet_len - offset_to_payload,
+                                  src_mac_addr,
+                                  dst_mac_addr,
+                                  dgram_size,
+                                  dgram_tag,
+                                  dgram_offset)
+        } else {
+            self.receive_single_packet(&packet, packet_len, src_mac_addr, dst_mac_addr)
+        }
+    }
+
+    fn receive_single_packet(&self,
+                             payload: &[u8],
+                             payload_len: usize,
+                             src_mac_addr: MacAddress,
+                             dst_mac_addr: MacAddress)
+                             -> (Option<&RxState<'a>>, ReturnCode) {
+        let rx_state = self.rx_states.iter().find(|state| !state.busy.get());
+        rx_state.map(|state| {
+                state.start_receive(src_mac_addr, dst_mac_addr, payload_len as u16, 0);
+                // The packet buffer should *always* be there; in particular,
+                // since this state is not busy, it must have the packet buffer.
+                // Otherwise, we are in an inconsistent state and can fail.
+                let mut packet = state.packet
+                    .take()
+                    .expect("Error: `packet` in RxState struct is `None` \
```

The final `expect` call is in the `receive_single_packet` function (line 810). Again, this buffer is owned by the `RxState` struct and if it is not there, we cannot receive any more frames. I had converted this buffer from a `TakeCell` to a `MapCell`, but this (again) touches on the "fail fast" debate, and whether to panic or return an error.